### DISCUSSION TITLE
Not unmounted as expected

### DISCUSSION THREAD

**Discussion by nogaten** [pmndrs/jotai#2658] [General]
I am using jotai 2.9.0.
In the following program, atom a is not unmounted as expected.

```typescript
const a = atom(0);
a.onMount = () => {
  console.log("a is mounted");
  return () => {
    console.log("a is unmounted");
  };
};

const b = atom(true);

const c = atom((get) => {
  if (get(b)) {
    console.log("mount a");
    get(a);
  } else {
    console.log("unmount a");
  }
});

const store = createStore();
store.sub(c, () => {});
setTimeout(() => {
  store.set(b, (state) => !state);
}, 1);
```

**Comment by dai-shi** [comment]
Hi, thanks for reporting.
Is this a regression in v2.9.0? Does it work differently in v2.8.4?

**Reply by nogaten** [reply]
It worked differently in my environment.

in v2.8.4
```
mount a
a is mounted
unmount a
a is unmounted
```

in v2.9.0
```
mount a
a is mounted
unmount a
```

**Reply by dai-shi** [reply]
That sounds like a bug.
Would you be able to create a PR with a failing test case?

**Reply by nogaten** [reply]
I haven't investigated the cause yet, but once I find out, I'll submit a PR.

**Reply by dai-shi** [reply]
sure.
even without you fixing it, the failing test should help (and someone else can tackle it.)
